### PR TITLE
 Add configurable activation function to MLP models #41 

### PR DESCRIPTION
**Pull request by JARACH-209**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue
Added activation functions 

https://github.com/facebookresearch/mbrl-lib/issues/41#issue-812559086

## How Has This Been Tested (if it applies)
Not yet !

## Checklist

- [ ] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](../../CONTRIBUTING.md) document and completed the CLA (see **CONTRIBUTING**).
- [ ] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by JARACH-209** [comment]
image

I am unable to resolve this pre-commit issue, if you can help!

File: mbrl/models/gaussian_mlp.py [hunk 1]
```diff
@@ -81,15 +81,18 @@ def __init__(
         deterministic: bool = False,
         propagation_method: Optional[str] = None,
         learn_logvar_bounds: bool = False,
+        activation = relu,
```
[the remark this answers is not in the document]
Comment (JARACH-209) [reply mbrl/models/gaussian_mlp.py]: I have used `activation_name` for better readability and distinction between activation class and name.

**Comment by JARACH-209** [comment]
@luisenp How about we just let the users pass the torch.nn.XXX activation function object itself. We wouldn't have to worry about handling the configurations like setting threshold values, lambda, etc.
`
activation = torch.nn.Threshold(0.5,10)
mlp = mbrl.models.GaussianMLP(..., activation_cls = activation)
`

File: mbrl/models/gaussian_mlp.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (luisenp) [reply mbrl/models/gaussian_mlp.py]: Agreed on `activation_name`. On the other hand, while we will certainly remove `use_silu` eventually, this is a change that breaks existing functionality, since our current `hydra` configurations won't work anymore. So, rather than removing it, we should emit a deprecation warning saying that this argument has now been replaced by `activation_name`. See [this](https://github.com/facebookresearch/mbrl-lib/blob/fa3f9b922db9b61d81cafb359546c5bb78440cd6/mbrl/util/replay_buffer.py#L538) for an example of how we have handled this in the past.

File: mbrl/models/gaussian_mlp.py [hunk 2]
```diff
@@ -77,19 +79,19 @@ def __init__(
...
         super().__init__(
             ensemble_size, device, propagation_method, deterministic=deterministic
         )
 
         self.in_size = in_size
         self.out_size = out_size
-
-        activation_cls = nn.SiLU if use_silu else nn.ReLU
+        if activation_cls is None:
+            activation_cls = nn.ReLU
```
Comment (luisenp) [inline mbrl/models/gaussian_mlp.py:94]: One issue with this is that, as far as I know, it won't work with `hydra` (see [config](https://github.com/facebookresearch/mbrl-lib/blob/master/mbrl/examples/conf/dynamics_model/gaussian_mlp.yaml)). If we want to pass the class directly, we should do something like [this config](https://github.com/facebookresearch/mbrl-lib/blob/master/mbrl/examples/conf/dynamics_model/basic_ensemble.yaml#L9), and instantiate like [this](https://github.com/facebookresearch/mbrl-lib/blob/fa3f9b922db9b61d81cafb359546c5bb78440cd6/mbrl/models/basic_ensemble.py#L74) in the code.

**Comment by luisenp** [comment]
Left a few more comments @JARACH-209, but I like your suggestions. Once we finalize the changes, we should also update the relevant configurations in `conf/examples`, and also the example [notebooks](https://github.com/facebookresearch/mbrl-lib/tree/fa3f9b922db9b61d81cafb359546c5bb78440cd6/notebooks), to use the new API. And finally, we should add a unit test in [this file](https://github.com/facebookresearch/mbrl-lib/blob/master/tests/core/test_models.py) to check that the model is created correctly.

**Comment by JARACH-209** [comment]
@luisenp passing the class seemed a bit tricky to me. However, see the recent commit as I hope this should suffice. I have replaced activation_cls() with configured activation functions because prior would only work for activation functions that do not take any arguments.

File: mbrl/models/gaussian_mlp.py [hunk 3]
```diff
@@ -56,14 +56,16 @@ class GaussianMLP(Ensemble):
...
+        use_silu (bool) (depricated, use activation_cls instead): if ``True``, hidden layers will use SiLU activations, otherwise
                          ReLU activations will be used. Defaults to ``False``.
         deterministic (bool): if ``True``, the model will be trained using MSE loss and no
             logvar prediction will be done. Defaults to ``False``.
         propagation_method (str, optional): the uncertainty propagation method to use (see
             above). Defaults to ``None``.
         learn_logvar_bounds (bool): if ``True``, the logvar bounds will be learned, otherwise
             they will be constant. Defaults to ``False``.
+        activation_name (str, optional): name of the activation function to use from torch.nn. Defaults to ``ReLU``.
+        activation_params (list, optinal): list of parameteres, if required, for the chosen activation as per the definition in torch.nn. Defaults to None. (e.g., activation_name = ``Threshold``, activation_params = [0.5,10] (threshold, value))
```
Comment (luisenp) [inline mbrl/models/gaussian_mlp.py:68]: Rather than pass the parameters, let's take advantage of Hydra here. All of the activation information (including the class name) can be passed as a dictionary and instantiated as described in [this](https://github.com/facebookresearch/mbrl-lib/pull/124#discussion_r697859674) comment, following the example of `BasicEnsemble`.

Comment (luisenp) [inline mbrl/models/gaussian_mlp.py:68]: This line is too long, max width should be 100 characters for comments. I'd even prefer 88, the default width for `black`, which we use for formatting.

File: mbrl/models/gaussian_mlp.py [hunk 4]
```diff
@@ -89,19 +93,27 @@ def __init__(
         self.in_size = in_size
         self.out_size = out_size
 
-        activation_cls = nn.SiLU if use_silu else nn.ReLU
+        if use_silu:
+            warnings.warn("use_silu is deprecated and will be removed.
+                          Use activation_name = 'SiLU' instead.")
+            activation_func = nn.SiLU()
+
+        if activation_name == "ReLU" and not use_silu:
```
Comment (luisenp) [inline mbrl/models/gaussian_mlp.py:101]: I think it's fine if the new mechanism is always used when if `use_silu = False`, so we can skip this special case for `nn.ReLU()` (as long as the default when nothing is specified is a relu activation).

File: mbrl/models/gaussian_mlp.py [hunk 5]
```diff
@@ -124,6 +136,12 @@ def create_linear_layer(l_in, l_out):
 
         self._propagation_indices: torch.Tensor = None
 
+    def _get_activation_func(self,activation_name, activation_params):
+        activation_func = getattr(nn,activation_name)
```
Comment (luisenp) [inline mbrl/models/gaussian_mlp.py:140]: There is a lint error here (no space after a comma). I got a message asking about some issues with pre-commit, but I couldn't see it in the repo. Were you able to solve them? Let me know if you have any questions about this.

File: mbrl/models/gaussian_mlp.py [hunk 6]
```diff
@@ -89,19 +93,27 @@ def __init__(
...
-        activation_cls = nn.SiLU if use_silu else nn.ReLU
+        if use_silu:
+            warnings.warn("use_silu is deprecated and will be removed.
+                          Use activation_name = 'SiLU' instead.")
+            activation_func = nn.SiLU()
+
+        if activation_name == "ReLU" and not use_silu:
+            activation_func = nn.ReLU()
+        else:
+            activation_func = self._get_activation_func(activation_name,activation_params)
```
Comment (luisenp) [inline mbrl/models/gaussian_mlp.py:104]: As mentioned above, since we are also including parameters, it's then better to replace this with `hydra.hydra.utils.instantiate(activation_fn_cfg)`, where `activation_fn_cfg` is a an optional dictionary with the target class and parameters. The default for this dictionary should be `None`, in which case the code defaults to `ReLU`.

**Review by luisenp — CHANGES_REQUESTED** [review]
Thanks for all the updates! I still have some comments, do let me know if anything is unclear.

File: mbrl/models/gaussian_mlp.py (hunk 2 above)
Comment (luisenp) [reply mbrl/models/gaussian_mlp.py]: Sorry, the above was probably confusing. What I meant is that we cannot pass custom objects directly due to `Hydra`. I left other follow up comments about using a config for the activation.

File: mbrl/models/gaussian_mlp.py (hunk 6 above)
Comment (JARACH-209) [reply mbrl/models/gaussian_mlp.py:104]: I don't fully understand the usage of hydra here but I will give it a shot !

File: mbrl/models/gaussian_mlp.py (hunk 5 above)
Comment (JARACH-209) [reply mbrl/models/gaussian_mlp.py:140]: Yes, I could solve the issues with pre-commit!.
image

File: mbrl/models/gaussian_mlp.py (hunk 6 above)
Comment (JARACH-209) [reply mbrl/models/gaussian_mlp.py:104]: @luisenp please review the changes I have made and let me know :)

Comment (JARACH-209) [reply mbrl/models/gaussian_mlp.py]: Once the changes are agreed upon, I'll go ahead with adding tests and updating the example notebook.

Comment (luisenp) [reply mbrl/models/gaussian_mlp.py]: Thanks a lot for the changes. I'm busy today, but I'm planning to look at this tomorrow or Wednesday. Thanks!

**Comment by luisenp** [comment]
@JARACH-209 did you delete this to make another PR from your master branch?
